### PR TITLE
Extract CRV filtering and enable usage in CallbackNLG

### PR DESCRIPTION
**Pull request by ancalita**
**Proposed changes**:
- Implements [ATO-1184](https://rasahq.atlassian.net/browse/ATO-1184)

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

[ATO-1184]: https://rasahq.atlassian.net/browse/ATO-1184?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: rasa/core/nlg/callback.py [hunk 1]
```diff
@@ -108,3 +114,25 @@ def validate_response(content: Optional[Dict[Text, Any]]) -> bool:
...
                 f"https://github.com/RasaHQ/rasa/blob/main/rasa/core/nlg/callback.py"
             )
+
+    @staticmethod
+    def fetch_response_id(
+        utter_action: Text,
+        tracker: DialogueStateTracker,
+        output_channel: Text,
+        domain_responses: Optional[Dict[Text, List[Dict[Text, Any]]]],
+    ) -> Text:
```
Comment (vcidst) [inline rasa/core/nlg/callback.py:124]: Can you please add a small docstring or a comment here?

File: rasa/core/nlg/generator.py [hunk 2]
```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
...
+    @staticmethod
+    def _matches_filled_slots(
+        filled_slots: Dict[Text, Any], response: Dict[Text, Any]
+    ) -> bool:
+        """Checks if the conditional response variation matches the filled slots."""
+        constraints = response.get(RESPONSE_CONDITION, [])
+        for constraint in constraints:
+            name = constraint["name"]
+            value = constraint["value"]
+            filled_slots_value = filled_slots.get(name)
```
Comment (vcidst) [inline rasa/core/nlg/generator.py:110]: I think its stylistically inconsistent to use `dict[key]` along with `dict.get(key)`. Can we use dict.get() in most cases?

File: rasa/core/nlg/generator.py [hunk 3]
```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
...
+            if isinstance(filled_slots_value, str) and isinstance(value, str):
+                if filled_slots_value.casefold() != value.casefold():
```
Comment (vcidst) [inline rasa/core/nlg/generator.py:112]: I didn't know about `str.casefold()` 🤔 good to know about this!

File: rasa/core/nlg/generator.py [hunk 4]
```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
...
+                    return False
+            elif filled_slots_value != value:
```
Comment (vcidst) [inline rasa/core/nlg/generator.py:114]: This is a strange if-else case, won't this only be executed if either `filled_slots_value` or `value` (aside: value is a vague variable name) are not strings? It looks odd to me because it looks like an `else` case for L112 `if filled_slots_value.casefold() != value.casefold():`

File: rasa/core/nlg/generator.py [hunk 5]
```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
...
+            filter(
+                lambda x: (
+                    x.get(RESPONSE_CONDITION)
+                    and self._matches_filled_slots(
+                        filled_slots=filled_slots, response=x
+                    )
+                ),
+                self.responses[utter_action],
+            )
+        )
```
Comment (vcidst) [inline rasa/core/nlg/generator.py:142]: clever code but its hard to read, can you please expand it a little bit? Or add comments that explain what the lamba function does

File: rasa/core/nlg/generator.py [hunk 6]
```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
...
+        conditional_no_channel = list(
+            filter(lambda x: (x.get(CHANNEL) is None), conditional_responses)
+        )
+        default_channel = list(
+            filter(lambda x: (x.get(CHANNEL) == output_channel), default_responses)
+        )
+        default_no_channel = list(
+            filter(lambda x: (x.get(CHANNEL) is None), default_responses)
+        )
```
Comment (vcidst) [inline rasa/core/nlg/generator.py:155]: Wouldn't it be more performant to iterate through the list `conditional_responses` and `default_responses` to filter them? Instead of iterating through them twice

File: rasa/core/nlg/generator.py [hunk 7]
```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
...
+
+        if conditional_channel:
+            return conditional_channel
```
Comment (vcidst) [inline rasa/core/nlg/generator.py:158]: If we move this check for empty list right after `conditional_channel` is created, that will save us the computation of other lists and help return early

File: rasa/core/nlg/generator.py [hunk 8]
```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
...
+
+        return default_no_channel
+
+    def get_response_variation_id(
+        self,
+        utter_action: Text,
+        tracker: DialogueStateTracker,
+        output_channel: Text,
+    ) -> Optional[Text]:
+        """Returns the first matched response variation ID that fits the channel, action and condition."""  # noqa: E501
```
Comment (vcidst) [inline rasa/core/nlg/generator.py:174]: I'm not sure why `noqa: E501` was required here. Can't we just split the docstring in multiple lines?

File: rasa/core/nlg/generator.py [hunk 9]
```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
...
+            response_variations: The response variations to validate.
+
+        Returns:
+            True if the response IDs are unique, False otherwise.
+        """
+        response_ids = set()
+        for response_variation in response_variations:
+            response_variation_id = response_variation.get("id")
+            if response_variation_id:
+                if response_variation_id in response_ids:
```
Comment (vcidst) [inline rasa/core/nlg/generator.py:201]: I think the condition `if response_variation_id in response_ids:` is sufficient and also checks for empty variable, what do you think?

File: rasa/core/nlg/response.py [hunk 10]
```diff
@@ -26,69 +26,6 @@ def __init__(self, responses: Dict[Text, List[Dict[Text, Any]]]) -> None:
...
-            filter(
-                lambda x: (
-                    x.get(RESPONSE_CONDITION)
-                    and self._matches_filled_slots(
-                        filled_slots=filled_slots, response=x
-                    )
-                ),
-                self.responses[utter_action],
-            )
-        )
```
Comment (vcidst) [inline rasa/core/nlg/response.py:66]: I see that you didn't write the clever code block! Its ok if you don't want to refactor it in `generator.py`. I hadn't seen this file until this and wasn't aware you've only re-arranged this block

Comment (ancalita) [reply rasa/core/nlg/response.py:66]: Yep, just added some context in [this reply](https://github.com/RasaHQ/rasa/pull/12575#discussion_r1246257045).
I think it would be really helpful if you could also look over the tests for the introduced functionality and think of any edge cases that I might have missed 🙏🏻

File: rasa/core/nlg/generator.py (hunk 2 above)
Comment (ancalita) [reply rasa/core/nlg/generator.py:110]: We always know that the constraints in a CRV will have those keys (name and value) because it's validated by the [YAML schema](https://github.com/RasaHQ/rasa/blob/6e10e42d052190577cafd9b22e79e367a60a5dc2/rasa/shared/nlu/training_data/schemas/responses.yml#L59), while filled_slots could have no value set for that slot, which is why it's needed to use `dict.get`. I also want to prevent adding checks for None to make mypy happy for those particular constraint keys. 

If you find it useful, I can add `required: True` to the keys in the yaml schema, this will definitely ensure that the keys are present.

File: rasa/core/nlg/generator.py (hunk 4 above)
Comment (ancalita) [reply rasa/core/nlg/generator.py:114]: This elif is required for all slots that have a different value type than strings, like numbers or booleans, so it will be executed in these particular instances.

File: rasa/core/nlg/generator.py (hunk 6 above)
Comment (ancalita) [reply rasa/core/nlg/generator.py:155]: As per my reply [here](https://github.com/RasaHQ/rasa/pull/12575/files#r1246257045), refactoring pre-existing code is not in scope.

File: rasa/core/nlg/generator.py (hunk 8 above)
Comment (ancalita) [reply rasa/core/nlg/generator.py:174]: I can try but I remember the linter complains if it doesn't encounter a `.` period char in the first line.

File: rasa/core/nlg/generator.py (hunk 9 above)
Comment (ancalita) [reply rasa/core/nlg/generator.py:201]: I'd combine the two actually by an AND to remove the indentation, but the first check is still required because we could be adding None values to the set in L208.

File: rasa/core/nlg/callback.py [hunk 11]
```diff
@@ -39,11 +39,11 @@ def nlg_request_format(
 ) -> Dict[Text, Any]:
     """Create the json body for the NLG json body for the request."""
     tracker_state = tracker.current_state(EventVerbosity.ALL)
-    response_ids = kwargs.pop(RESPONSE_ID_KEY, [])
+    response_id = kwargs.pop("response_id", "")
 
     return {
         "response": utter_action,
-        "ids": response_ids,
+        "id": response_id,
```
[the remark this answers is not in the document]
Comment (ancalita) [reply rasa/core/nlg/callback.py:46]: Good point, I made a decision for an empty string because I thought that I can't send a None value in a POST request body, but maybe I got confused by some other testing issues encountered while I was working on the implementation. I will double-check now if None is possible and also ask CSE if they prefer one or the other.

File: rasa/core/nlg/generator.py (hunk 7 above)
Comment (ancalita) [reply rasa/core/nlg/generator.py:158]: Actually `filter` function (along with map and reduce) is super fast and performance gain is negligible. I also don't want to tinker with something that works atm and no performance complaints have been registered.

File: rasa/core/nlg/callback.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (vcidst) [reply rasa/core/nlg/callback.py:46]: Sounds good!

File: rasa/core/nlg/generator.py (hunk 2 above)
Comment (vcidst) [reply rasa/core/nlg/generator.py:110]: Sounds good, I don't think adding `required: True` is needed

File: rasa/core/nlg/generator.py (hunk 4 above)
Comment (vcidst) [reply rasa/core/nlg/generator.py:114]: can you please add a small comment here to make this clear?

File: rasa/core/nlg/response.py (hunk 10 above)
Comment (vcidst) [reply rasa/core/nlg/response.py:66]: Yes, I couldn't complete the review yesterday. I'm going to take a look at the rest of the files in the PR now

File: rasa/core/nlg/generator.py (hunk 4 above)
Comment (ancalita) [reply rasa/core/nlg/generator.py:117]: Addressed: https://github.com/RasaHQ/rasa/pull/12575/commits/27e98159b985b91c4a2204fc18cc089e54d75efb

File: rasa/core/nlg/callback.py (hunk 1 above)
Comment (ancalita) [reply rasa/core/nlg/callback.py:124]: Addressed: https://github.com/RasaHQ/rasa/pull/12575/commits/27e98159b985b91c4a2204fc18cc089e54d75efb

File: rasa/core/nlg/generator.py (hunk 9 above)
Comment (ancalita) [reply rasa/core/nlg/generator.py]: Addressed: https://github.com/RasaHQ/rasa/pull/12575/commits/27e98159b985b91c4a2204fc18cc089e54d75efb

File: rasa/core/nlg/callback.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ancalita) [reply rasa/core/nlg/callback.py:46]: None seems to be possible, so I've pinged CSE on the ticket to enquire which option they prefer. In the meantime, I'll re-request review 😄

**Review by vcidst — APPROVED** [review]
LGTM 🚀

File: changelog/12663.doc.md [hunk 12]
```diff
@@ -1 +1 @@
-Updated commands with square brackets e.g (`pip install rasa[spacy]`) to use quotes (`pip install 'rasa[spacy]'`) for compatibility with zsh in docs.
\ No newline at end of file
+Updated commands with square brackets e.g (`pip install rasa[spacy]`) to use quotes (`pip install 'rasa[spacy]'`) for compatibility with zsh in docs.
```
Comment (ancalita) [inline changelog/12663.doc.md:1]: Had to amend this incoming changelog from merging 3.6.x into my branch because of [code quality job failure](https://github.com/RasaHQ/rasa/actions/runs/5532780783/jobs/10095367732#step:15:23).

Comment (vcidst) [reply changelog/12663.doc.md:1]: That's so strange, how did the PR pass with that?

Comment (ancalita) [reply changelog/12663.doc.md:1]: No idea, added this code: https://github.com/RasaHQ/rasa/pull/12575/commits/efada8b2a3f1d6663410afd4f3b29b12ca4969a1, please re-review 🙏🏻
